RIOT-OS/RIOT
# riotboot: implement serial flasher

### Contribution description

I got annoyed by boards that require wiring up a programmer and that are finicky to flash.
Therefore I added a simple protocol to `riotboot` that allows to erase sectors / write data over a serial connection.

There are only four commands:

 - erase sector
 - get sector of address
 - write data to address
 - boot

This comes with a custom flash tool (`riotboot_serial`) that is used to transfer the firmware onto a board using riotboot.
So far, this is only wired up for the sam0 family (and comes with a patch that fixes a hard fault on samd5x when writing small chunks).

**edit** also added efm32 and nrf52, successfully flashed boards using riotboot serial with those platforms.

The flash tool is rather simple and does not do proper error handling yet (it expects the user to just interrupt it if it gets stuck).
The lines from the `.hex` file are converted to binary 

I considered sending raw lines from the `.hex` file directly, but decided against so the bootloader doesn't have to parse text.
Also sending binary over the wire should be faster.

If the sync protocol makes sense / the 'double sync' is necessary is up for debate, also batching multiple lines of the hex file would probably be faster.

There is also no support for doing anything with riotboot headers yet.
If they are part of the hex file and can be handled transparently, it should just work, but I did not test this.

### Testing procedure

For the riotboot bootloader, enable the `riotboot_serial` and disable `riotboot_slot`.
It is possible to have both modules active at the same time, but then riotboot will also check if the slot has a valid header, which makes testing/firmware generation more complicated.

With this configuration, 'raw' images (plain RIOT firmware) will boot.

```patch
--- a/bootloaders/riotboot/Makefile
+++ b/bootloaders/riotboot/Makefile
@@ -21,10 +21,10 @@ DISABLE_MODULE += auto_init auto_init_%
 USEMODULE += stdio_null
 
 # Include riotboot flash partition functionality
-USEMODULE += riotboot_slot
+# USEMODULE += riotboot_slot
 
 # Include serial bootloader functionality
-# USEMODULE += riotboot_serial
+USEMODULE += riotboot_serial
```

Now you should be able to flash any RIOT application using `PROGRAMMER=riotboot_serial`. Start the flashing procedure, then reset the board for the flasher to sync up with the bootloader.
You can also enable the `riotboot_reset` module which gives you the `bootloader` command to reboot to the bootloader from the shell.

### Issues/PRs references
